File: examples/py/04_variants_sweep.py
"""Example 04: Variants (Parameter Sweep)

Demonstrates running multiple nodes with the same structure but different parameters
"""

import sys
sys.path.insert(0, '/home/runner/work/graph-sp/graph-sp/examples/py')

from benchmark_utils import Benchmark, print_header, print_section
import dagex


def data_source(_inputs):
    """Generate initial data."""
    return {"base": 10}


def make_multiplier(factor):
    """Factory function to create multiplier variants."""
    def multiplier(inputs):
        value = inputs.get("x", 0)
        return {"result": value * factor}
    return multiplier


def main():
    print_header("Example 04: Variants (Parameter Sweep)")
    
    print("📖 Story:")
    print("   Variants let you create many nodes with the same structure but")
    print("   different captured parameters. The graph will attach them to the")
    print("   same frontier and execute them at the same level when possible.")
    print("   This is perfect for hyperparameter sweeps or A/B testing.\n")
    
    print_section("Building the Graph")
    
    graph = dagex.Graph()
    
    # Add source
    graph.add(
        data_source,
        label="DataSource",
        inputs=None,
        outputs=[("base", "x")]
    )
    
    # Add variants with different multipliers
    factors = [2, 3, 5, 7]
    variant_nodes = [make_multiplier(f) for f in factors]
    
    graph.variants(
        variant_nodes,
        label="Multiplier",
        inputs=[("x", "x")],
        outputs=[("result", "results")]
    )
    
    dag = graph.build()
    
    print_section("Mermaid Diagram")
    print(dag.to_mermaid())
    
    print_section("ASCII Visualization")
    print("                Multiplier(×2)")
    print("               /")
    print("              |  Multiplier(×3)")
    print("  DataSource  <")
    print("              |  Multiplier(×5)")
    print("               \\")
    print("                Multiplier(×7)")
    
    print_section("Sequential Execution (parallel=False)")
    
    with Benchmark("Sequential execution") as bench_seq:
        context_seq = dag.execute(parallel=False)
    
    bench_seq.print_result()
    result_seq = bench_seq.result
    
    print_section("Parallel Execution (parallel=True)")
    
    with Benchmark("Parallel execution") as bench_par:
        context_par = dag.execute(parallel=True, max_threads=4)
    
    bench_par.print_result()
    result_par = bench_par.result
    
    print_section("Results")
    
    print("📊 Base value: 10")
    
    print("\nSequential execution:")
    print(f"  Time: {result_seq.duration_ms:.3f}ms")
    
    print("\nParallel execution:")
    print(f"  Time: {result_par.duration_ms:.3f}ms")
    
    # Show detailed variant outputs
    print("\nDetailed variant outputs:")
    for i, factor in enumerate(factors):
        expected = 10 * factor
        print(f"  Variant {i} (×{factor}): {expected}")
    
    print(f"\n✅ All {len(factors)} variants executed successfully!")
    
    print()


if __name__ == "__main__":
    main()

File: examples/rs/04_variants_sweep.rs
// Example 04: Variants (Parameter Sweep)
// Demonstrates running multiple nodes with the same structure but different parameters

mod benchmark_utils;

use dagex::{Graph, GraphData};
use std::collections::HashMap;
use std::thread;
use std::time::Duration;
use benchmark_utils::{Benchmark, print_header, print_section};

fn data_source(_inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let mut outputs = HashMap::new();
    outputs.insert("base".to_string(), GraphData::int(10));
    outputs
}

// Factory function to create multiplier variants
fn make_multiplier(factor: i64) -> impl Fn(&HashMap<String, GraphData>) -> HashMap<String, GraphData> + Send + Sync + 'static {
    move |inputs: &HashMap<String, GraphData>| -> HashMap<String, GraphData> {
        let value = inputs.get("x").and_then(|d| d.as_int()).unwrap_or(0);
        
        // Simulate I/O-bound work (file read, network call, database query, etc.)
        // that benefits from parallelization
        thread::sleep(Duration::from_millis(150));
        
        let mut outputs = HashMap::new();
        outputs.insert("result".to_string(), GraphData::int(value * factor));
        outputs
    }
}

fn main() {
    print_header("Example 04: Variants (Parameter Sweep)");
    
    println!("📖 Story:");
    println!("   Variants let you create many nodes with the same structure but");
    println!("   different captured parameters. The graph will attach them to the");
    println!("   same frontier and execute them at the same level when possible.");
    println!("   This is perfect for hyperparameter sweeps or A/B testing.\n");
    
    print_section("Building the Graph");
    
    let mut graph = Graph::new();
    
    // Add source
    graph.add(
        data_source,
        Some("DataSource"),
        None,
        Some(vec![("base", "x")])
    );
    
    // Add variants with different multipliers
    let factors = vec![2, 3, 5, 7];
    let variant_nodes: Vec<_> = factors
        .iter()
        .map(|&f| make_multiplier(f))
        .collect();
    
    graph.variants(
        variant_nodes,
        Some("Multiplier"),
        Some(vec![("x", "x")]),
        Some(vec![("result", "results")])
    );
    
    let dag = graph.build();
    
    print_section("Mermaid Diagram");
    println!("{}", dag.to_mermaid());
    
    print_section("ASCII Visualization");
    println!("                Multiplier(×2)");
    println!("               /");
    println!("              |  Multiplier(×3)");
    println!("  DataSource  <");
    println!("              |  Multiplier(×5)");
    println!("               \\");
    println!("                Multiplier(×7)");
    
    print_section("Sequential Execution (parallel=false)");
    
    let bench = Benchmark::start("Sequential execution");
    let _result_seq = dag.execute_detailed(false, None);
    let bench_result_seq = bench.finish_and_print();
    
    print_section("Parallel Execution (parallel=true)");
    
    let bench = Benchmark::start("Parallel execution");
    let _result_par = dag.execute_detailed(true, Some(4));
    let bench_result_par = bench.finish_and_print();
    
    print_section("Results");
    
    println!("📊 Base value: 10");
    
    println!("\nSequential execution:");
    println!("  Time: {:.3}ms", bench_result_seq.duration_ms);
    
    println!("\nParallel execution:");
    println!("  Time: {:.3}ms", bench_result_par.duration_ms);
    
    // Show detailed node outputs if available
    println!("\nDetailed variant outputs:");
    for (i, factor) in factors.iter().enumerate() {
        let expected = 10 * factor;
        println!("  Variant {} (×{}): {}", i, factor, expected);
    }
    
    println!("\n✅ All {} variants executed successfully!", factors.len());
    
    println!();
}
